File: built-in plugins dsp/mb-comp-sidechain.cpp
/**
 * MB Sidechain Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Pumping sidechain compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_SIDECHAIN_H
#define MB_COMP_SIDECHAIN_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompSidechain : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-sidechain";
    static constexpr const char* PLUGIN_NAME    = "MB Sidechain Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -30f;  // range [-60, 0]
    float ratio = 10f;  // range [1, 20]
    float attack = 1f;  // range [0.1, 20]
    float release = 150f;  // range [20, 500]
    };

    MbCompSidechain() = default;
    ~MbCompSidechain() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -60f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 20f);
        params.attack = std::clamp(params.attack, 0.1f, 20f);
        params.release = std::clamp(params.release, 20f, 500f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Sidechain Comp
        return input;
    }
};

#endif // MB_COMP_SIDECHAIN_H

File: built-in plugins dsp/mb-comp-sidechain.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCompSidechainPlugin: PluginDefinition = { id: 'mb-comp-sidechain', slug: 'mb-comp-sidechain', name: 'MB Sidechain Comp', category: 'effect', type: 'compressor', version: '1.0.0', description: 'Pumping sidechain compressor', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -30, minValue: -60, maxValue: 0, automatable: true }, { id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 10, minValue: 1, maxValue: 20, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 1, minValue: 0.1, maxValue: 20, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 150, minValue: 20, maxValue: 500, automatable: true }], defaultPreset: { threshold: -30, ratio: 10, attack: 1, release: 150 } };

export default MbCompSidechainPlugin;
